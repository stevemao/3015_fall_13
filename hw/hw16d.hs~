import System.IO
import Data.Char

main = chat [] 

chat xs = 
    do putStrLn $ "queue so far = " ++ (show xs) 
       putStr $ ">  "
       hFlush stdout
       i <- getLine
       case i of 
          "?" -> do putStrLn $ "commands are: \"dequeue\", \"quit\" or, to queue up a number type the number."
                    chat xs
          "dequeue" ->  return () -- fix this case
          "quit" -> return ()
          "" -> return ()
          _ -> return () -- fix this case