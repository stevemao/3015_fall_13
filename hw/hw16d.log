This is pdfeTeX, Version 3.141592-1.21a-2.2 (Web2C 7.5.4) (format=pdflatex 2011.9.13)  28 OCT 2011 17:44
entering extended mode
**hw16d
(./hw16d.tex
LaTeX2e <2003/12/01>
Babel <v3.8d> and hyphenation patterns for american, french, german, ngerman, b
ahasa, basque, bulgarian, catalan, croatian, czech, danish, dutch, esperanto, e
stonian, finnish, greek, icelandic, irish, italian, latin, magyar, norsk, polis
h, portuges, romanian, russian, serbian, slovak, slovene, spanish, swedish, tur
kish, ukrainian, nohyphenation, loaded.
(/usr/share/texmf/tex/latex/base/article.cls
Document Class: article 2004/02/16 v1.4f Standard LaTeX document class
(/usr/share/texmf/tex/latex/base/size11.clo
File: size11.clo 2004/02/16 v1.4f Standard LaTeX file (size option)
)
\c@part=\count79
\c@section=\count80
\c@subsection=\count81
\c@subsubsection=\count82
\c@paragraph=\count83
\c@subparagraph=\count84
\c@figure=\count85
\c@table=\count86
\abovecaptionskip=\skip41
\belowcaptionskip=\skip42
\bibindent=\dimen102
) (./qtree.sty
Package: 2008/12/12 v.3.1bQtree: tree-drawing for linguistics
Package: qtree 2008/12/12 v.3.1bQtree: tree-drawing for linguistics
(/usr/share/texmf/tex/latex/pict2e/pict2e.sty
Package: pict2e 2004/08/06 v0.2q Improved picture commands (HjG,RN)

(/usr/share/texmf/tex/latex/pict2e/pict2e.cfg
File: pict2e.cfg 2004/06/25 v0.1r pict2e configuration for teTeX/TeXLive
)
Package pict2e Info: Driver file: pdftex.def on input line 92.

(/usr/share/texmf/tex/latex/graphics/pdftex.def
File: pdftex.def 2002/06/19 v0.03k graphics/color for pdftex
\Gread@gobject=\count87
)
Package pict2e Info: Driver file for pict2e: p2e-pdftex.def on input line 94.

(/usr/share/texmf/tex/latex/pict2e/p2e-pdftex.def
File: p2e-pdftex.def 2004/06/25 v0.1r Driver-dependant file (RN,HjG)
)
(/usr/share/texmf/tex/latex/graphics/trig.sty
Package: trig 1999/03/16 v1.09 sin cos tan (DPC)
)
\pIIe@GRAPH=\toks14
)
\c@treecount=\count88
\c@branchcount=\count89
\parentbox=\box26
\treebox=\box27
\treeboxone=\box28
\treeboxtwo=\box29
\treeboxthree=\box30
\treeboxfour=\box31
\treeboxfive=\box32
\treeboxsix=\box33
\treeboxseven=\box34
\treeboxeight=\box35
\treeboxnine=\box36
\treeboxten=\box37
\treeboxeleven=\box38
\treeboxtwelve=\box39
\treeboxthirteen=\box40
\treeboxfourteen=\box41
\treeboxfifteen=\box42
\treeboxsixteen=\box43
\treeboxseventeen=\box44
\treeboxeighteen=\box45
\treeboxnineteen=\box46
\treeboxtwenty=\box47
\treeoffsetone=\skip43
\treeoffsettwo=\skip44
\treeoffsetthree=\skip45
\treeoffsetfour=\skip46
\treeoffsetfive=\skip47
\treeoffsetsix=\skip48
\treeoffsetseven=\skip49
\treeoffseteight=\skip50
\treeoffsetnine=\skip51
\treeoffsetten=\skip52
\treeoffseteleven=\skip53
\treeoffsettwelve=\skip54
\treeoffsetthirteen=\skip55
\treeoffsetfourteen=\skip56
\treeoffsetfifteen=\skip57
\treeoffsetsixteen=\skip58
\treeoffsetseventeen=\skip59
\treeoffseteighteen=\skip60
\treeoffsetnineteen=\skip61
\treeoffsettwenty=\skip62
\treeshiftone=\skip63
\treeshifttwo=\skip64
\treeshiftthree=\skip65
\treeshiftfour=\skip66
\treeshiftfive=\skip67
\treeshiftsix=\skip68
\treeshiftseven=\skip69
\treeshifteight=\skip70
\treeshiftnine=\skip71
\treeshiftten=\skip72
\treeshifteleven=\skip73
\treeshifttwelve=\skip74
\treeshiftthirteen=\skip75
\treeshiftfourteen=\skip76
\treeshiftfifteen=\skip77
\treeshiftsixteen=\skip78
\treeshiftseventeen=\skip79
\treeshifteighteen=\skip80
\treeshiftnineteen=\skip81
\treeshifttwenty=\skip82
\treewidthone=\skip83
\treewidthtwo=\skip84
\treewidththree=\skip85
\treewidthfour=\skip86
\treewidthfive=\skip87
\treewidthsix=\skip88
\treewidthseven=\skip89
\treewidtheight=\skip90
\treewidthnine=\skip91
\treewidthten=\skip92
\treewidtheleven=\skip93
\treewidthtwelve=\skip94
\treewidththirteen=\skip95
\treewidthfourteen=\skip96
\treewidthfifteen=\skip97
\treewidthsixteen=\skip98
\treewidthseventeen=\skip99
\treewidtheighteen=\skip100
\treewidthnineteen=\skip101
\treewidthtwenty=\skip102
\daughteroffsetone=\skip103
\daughteroffsettwo=\skip104
\daughteroffsetthree=\skip105
\daughteroffsetfour=\skip106
\branchwidthone=\skip107
\branchwidthtwo=\skip108
\branchwidththree=\skip109
\branchwidthfour=\skip110
\parentoffset=\skip111
\treeoffset=\skip112
\daughteroffset=\skip113
\branchwidth=\skip114
\parentwidth=\skip115
\treewidth=\skip116
\qta=\toks15
\qtb=\toks16
\nbranches=\count90
\qroofx=\count91
\qroofy=\count92
\@qrscratchbox=\box48
)
\@input{header.aux}
\openout2 = `header.aux'.

(./header.tex (/usr/share/texmf/tex/latex/amsfonts/amssymb.sty
Package: amssymb 2002/01/22 v2.2d

(/usr/share/texmf/tex/latex/amsfonts/amsfonts.sty
Package: amsfonts 2001/10/25 v2.2f
\@emptytoks=\toks17
\symAMSa=\mathgroup4
\symAMSb=\mathgroup5
LaTeX Font Info:    Overwriting math alphabet `\mathfrak' in version `bold'
(Font)                  U/euf/m/n --> U/euf/b/n on input line 132.
)) (./remark.sty)
(/home/faculty/jlc/papers/tex-stuff/bussproofs.sty
Proof Tree (bussproofs) style macros. Version 0.6c.
\theLevel=\count93
\myMaxLevel=\count94
\myBoxA=\box49
\myBoxB=\box50
\myBoxC=\box51
\myBoxD=\box52
\myBoxLL=\box53
\myBoxRL=\box54
\thisAboveSkip=\dimen103
\thisBelowSkip=\dimen104
\newScoreStart=\dimen105
\newScoreEnd=\dimen106
\newCenter=\dimen107
\displace=\dimen108
\leftLowerAmt=\dimen109
\rightLowerAmt=\dimen110
\scoreHeight=\dimen111
\scoreDepth=\dimen112
)
\c@theorem=\count95
\c@Rule=\count96
\c@conjecture=\count97
\c@corollary=\count98
\c@example=\count99
\c@fact=\count100
\c@lemma=\count101
\c@definition=\count102
\c@claim=\count103
\c@remark=\count104
\c@slogan=\count105
\c@axiom=\count106
\c@note=\count107
\c@problem=\count108
\c@exercise=\count109
) (./hw16d.aux)
\openout1 = `hw16d.aux'.

LaTeX Font Info:    Checking defaults for OML/cmm/m/it on input line 6.
LaTeX Font Info:    ... okay on input line 6.
LaTeX Font Info:    Checking defaults for T1/cmr/m/n on input line 6.
LaTeX Font Info:    ... okay on input line 6.
LaTeX Font Info:    Checking defaults for OT1/cmr/m/n on input line 6.
LaTeX Font Info:    ... okay on input line 6.
LaTeX Font Info:    Checking defaults for OMS/cmsy/m/n on input line 6.
LaTeX Font Info:    ... okay on input line 6.
LaTeX Font Info:    Checking defaults for OMX/cmex/m/n on input line 6.
LaTeX Font Info:    ... okay on input line 6.
LaTeX Font Info:    Checking defaults for U/cmr/m/n on input line 6.
LaTeX Font Info:    ... okay on input line 6.
 (/usr/share/texmf/tex/context/base/supp-pdf.tex
(/usr/share/texmf/tex/context/base/supp-mis.tex
loading : Context Support Macros / Miscellaneous (2004.10.26)
\protectiondepth=\count110
\scratchcounter=\count111
\scratchtoks=\toks18
\scratchdimen=\dimen113
\scratchskip=\skip117
\scratchmuskip=\muskip10
\scratchbox=\box55
\scratchread=\read1
\scratchwrite=\write3
\zeropoint=\dimen114
\onepoint=\dimen115
\onebasepoint=\dimen116
\minusone=\count112
\thousandpoint=\dimen117
\onerealpoint=\dimen118
\emptytoks=\toks19
\nextbox=\box56
\nextdepth=\dimen119
\everyline=\toks20
\!!counta=\count113
\!!countb=\count114
\recursecounter=\count115
)
loading : Context Support Macros / PDF (2004.03.26)
\nofMPsegments=\count116
\nofMParguments=\count117
\MPscratchCnt=\count118
\MPscratchDim=\dimen120
\MPnumerator=\count119
\everyMPtoPDFconversion=\toks21
)
Underfull \hbox (badness 10000) in paragraph at lines 7--8

 []

LaTeX Font Info:    Try loading font information for U+msa on input line 19.
(/usr/share/texmf/tex/latex/amsfonts/umsa.fd
File: umsa.fd 2002/01/19 v2.2g AMS font definitions
)
LaTeX Font Info:    Try loading font information for U+msb on input line 19.

(/usr/share/texmf/tex/latex/amsfonts/umsb.fd
File: umsb.fd 2002/01/19 v2.2g AMS font definitions
)
No file hw16d.bbl.
[1



{/usr/share/texmf-var/fonts/map/pdftex/updmap/pdftex.map}] (./hw16d.aux) ) 
Here is how much of TeX's memory you used:
 1672 strings out of 94500
 18659 string characters out of 1176766
 71067 words of memory out of 1000000
 4857 multiletter control sequences out of 10000+50000
 9472 words of font info for 36 fonts, out of 500000 for 2000
 580 hyphenation exceptions out of 1000
 27i,7n,26p,255b,155s stack positions out of 1500i,500n,5000p,200000b,5000s
PDF statistics:
 25 PDF objects out of 300000
 0 named destinations out of 131072
 1 words of extra memory for PDF output out of 65536
</usr
/share/texmf/fonts/type1/bluesky/cm/cmsy10.pfb></usr/share/texmf/fonts/type1/bl
uesky/cm/cmmi10.pfb></usr/share/texmf/fonts/type1/bluesky/cm/cmsl10.pfb></usr/s
hare/texmf/fonts/type1/bluesky/cm/cmti10.pfb></usr/share/texmf/fonts/type1/blue
sky/cm/cmtt10.pfb></usr/share/texmf/fonts/type1/bluesky/cm/cmr10.pfb></usr/shar
e/texmf/fonts/type1/bluesky/cm/cmbx10.pfb>
Output written on hw16d.pdf (1 page, 48395 bytes).
