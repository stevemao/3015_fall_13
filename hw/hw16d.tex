\documentclass[11pt]{article}
\usepackage{qtree}
\include{header}


\begin{document}
\homework{16}{1 November}

\exercise{
Read chapter 9 of LYAHFGG. Read Bird pp. 319-322.  Real World Haskell also has some good stuff on I/O (see {\tt{http://book.realworldhaskell.org/read/io.html}}).
}

\exercise{
In this homework you will do some interactive IO.  You are to write an
interactive simulator for a queue of numbers.

You can represent the queue itself using a list. When an item is queued up add it to the right end of th list. When an item is dequeued, just take it off the left end of the list.  The commands your program should respond to are :\vspace{.125in}\\
\begin{centering}
\begin{tabular}{rcl}
{"?"$\;\;$} & &When this command is entered, display a message describing the commands.\\
{k$\;\;$} & & If k is a number, queue it up. \\
{"dequeue"$\;\;$} & & remove an item from the head of the queue.  \\
{"quit"$\;\;$} & &  exit the interactive loop.\\
{""$\;\;$} & &  exit the interactive loop on the empty string.\\
{\_\_$\;\;$} & &  For any other input suggest the user try the "?" command and to try again.\\
\end{tabular}
\end{centering}

\ \\Here's the basic structure of the program.

\begin{program**}
\>  import System.IO \\
\>  import Data.Char \vspace{.125in}\\
\>  main = chat []  \vspace{.125in}\\
\>  chat xs =  \\
\>      do putStrLn \$ "queue so far = " ++ (show xs)  \\
\>          putStr "$>$  " \\
\>          hFlush stdout \\
\>          i $\leftarrow$ getLine \\
\>          case i of  \\
\>            "?" $\rightarrow$ do putStrLn \$ \\
\>                         "commands are: $\backslash$"dequeue$\backslash$", $\backslash$"quit$\backslash$" or, to queue up a number, enter it." \\
\>                        chat xs \\
\>            "dequeue" $\rightarrow$  ... \\
\>            "quit" $\rightarrow$ return () \\
\>            "" $\rightarrow$ return () \\
\>            \_ $\rightarrow$ if (all isDigit i) then  \\
\>                   ... \\
\>                 else \\
\>                   ...
\end{program**}

}

\exercise{
For extra credit. Add  "save" and  "load" commands that allow the current state of the queue to be saved to a file and restored from a file.
}

\bibliographystyle{plain}
\bibliography{./homework}

\end{document}


% Local Variables:
% mode:latex
% comment-column:0
% comment-start: "\\>  "
% comment-end: "\\\\"
% compile-command: "pdflatex hw16d"
% fill-column:79
% End:




