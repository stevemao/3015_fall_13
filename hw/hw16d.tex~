\documentclass[11pt]{article}
\usepackage{qtree}
\include{header}


\begin{document}
\homework{16}{1 November}

In this homework you will do some IO.  You are to write an interactive simulator for a queue of numbers.

You can represent the queue itself using a list. When an item is queued up add it to the right end of th list. When an item is dequeued, just take it off the left end of the list.  The commands your program should respond to are :
\begin{description}
\item{``?''} When this command is entered, display a message describing the commands.
\item{k} If k is a number, queue it up.
\item{``dequeue''} remove an item from the head of the queue. 
\item{``quit''} exit the interactive loop.
\item{\_} For any other input suggest the user try the ``?'' command and to try again.
\end{description}


\exercise{Write a function {\tt{inv::Btree a $\rightarrow$ Rose a }} that implements the inverse mapping.}

\bibliographystyle{plain}
\bibliography{./homework}

\end{document}


% Local Variables:
% mode:latex
% comment-column:0
% comment-start: "\\>  "
% comment-end: "\\\\"
% compile-command: "pdflatex hw16c"
% fill-column:79
% End:




