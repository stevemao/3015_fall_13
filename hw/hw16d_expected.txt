
alpamayo 15% ./iQueue
./iQueue
queue so far = []
>  1
1
queue so far = [1]
>  2
2
queue so far = [1,2]
>  3
3
queue so far = [1,2,3]
>  dequeue
dequeue
queue so far = [2,3]
>  dequeue
dequeue
queue so far = [3]
>  dequeue
dequeue
queue so far = []
>  dequeue
dequeue
you can't dequeue the empty queue. Try again.
queue so far = []
>  1
1
queue so far = [1]
>  2
2
queue so far = [1,2]
>  3
3
queue so far = [1,2,3]
>  4
4
queue so far = [1,2,3,4]
>  5
5
queue so far = [1,2,3,4,5]
>  dequeue
dequeue
queue so far = [2,3,4,5]
>  dequeue
dequeue
queue so far = [3,4,5]
>  dequeue
dequeue
queue so far = [4,5]
>  dequeue
dequeue
queue so far = [5]
>  dequeue
dequeue
queue so far = []
>  dequeue
dequeue
you can't dequeue the empty queue. Try again.
queue so far = []
>  ?
?
commands are: "dequeue", "quit" or, to queue up a number type the number.
queue so far = []
>  